### ISSUE TITLE
x/tools/gopls: support autocompletion of unimported packages

### ISSUE THREAD

**Issue by stamblerre** [golang/go#31906]
See for https://github.com/Microsoft/vscode-go/issues/2484 for context.

**Comment by ramya-rao-a** [comment]
When this does get into `gopls` please send a PR or log an issue to the vscode Go plugin so that we can disable the similar feature (in non module mode) there to avoid duplicate suggestions

**Comment by zchee** [comment]
@stamblerre
Just a question.
Is this issue means only of stdlib like gocode?
Or, all of packages in`go.mod`?

**Comment by stamblerre** [comment]
It would be all available packages, not just stdlib.

**Comment by zchee** [comment]
@stamblerre 
great.

Are you doing now developing this feature?
I do not know what you remember me, I maintained nsf/gocode for a while. If you not yet develop this feature, I would want to help you.

**Comment by stamblerre** [comment]
I am not actively working on it yet, so you're absolutely welcome to get started on it.

**Comment by jackielii** [comment]
@stamblerre @saibing I found out that this is a difference between https://github.com/saibing/tools & https://github.com/golang/tools/tree/

**Comment by stamblerre** [comment]
From https://github.com/microsoft/vscode-go/issues/2575:

There are really two components to this issue. For example (in both cases, `time` is not imported):
1) Typing "`ti<>`" and getting package `time` as part of the autocompletion results.
2) Typing "`time.<>`" and getting autocompletions for the members of package `time`.

**Comment by heschik** [comment]
At master, gopls now supports completing all unimported packages. There may still be performance issues, but the functionality exists.

**Comment by islishude** [comment]
It works now!

![屏幕快照 2019-10-22 上午9 21 46](https://user-images.githubusercontent.com/24730006/67254074-9f430580-f4ad-11e9-8f62-1c61961f36b7.png)

**Comment by chengyayu** [comment]
> It works now!
> 
> ![屏幕快照 2019-10-22 上午9 21 46](https://user-images.githubusercontent.com/24730006/67254074-9f430580-f4ad-11e9-8f62-1c61961f36b7.png)

does it works for ```go modules``` ```gpls```? which version is it ?

**Comment by chengyayu** [comment]
**versions**

- vscode : 1.39.2
- vscode-go: 0.11.7

**my settings**

```json
 // go settings
    "go.buildOnSave": "off",
    "go.lintTool":"golangci-lint",
    "go.lintFlags": [
        "--fast"
    ],
    "go.useLanguageServer": true,
    "go.languageServerExperimentalFeatures": {
        "documentLink": false
    },
    "[go]": {
        "editor.snippetSuggestions": "none",
        "editor.formatOnSave": true,
        "editor.codeActionsOnSave": {
            "source.organizeImports": true
        }
    },
    "gopls": {
        "usePlaceholders": true, // add parameter placeholders when completing a function
        "completeUnimported": true, // autocomplete unimported packages
        "watchFileChanges": true,  // watch file changes outside of the editor
        "deepCompletion": true,     // enable deep completion
    },
```

when i type this codes below

```go
package main

func main() {
	time.No
}
```

**gopls output error**

```
[Info  - 5:23:18 PM] 2019/10/22 17:23:18 no completions found
	At = {3 8}
	Failure = cannot resolve time
```

**Comment by stamblerre** [comment]
We don't yet have the exports loaded for the unimported packages, so it won't complete if you type `grpc.`. However, if you begin typing `grpc`, you will see the completions suggestions pop up, and then when you select one and hit enter, the corresponding import will be added so you can then add the period and get further completions from the package.

**Comment by chengyayu** [comment]
>https://github.com/golang/go/issues/31906#issuecomment-545056493

Sorry No suggestions pop up

**Comment by stamblerre** [comment]
@chengyayu: Are you using `gopls` at master? To install it, you have to clone the repo, cd into the `gopls` directory and run `go install`.

**Comment by chengyayu** [comment]
> @chengyayu: Are you using `gopls` at master? To install it, you have to clone the repo, cd into the `gopls` directory and run `go install`.

yes! it works! thanks for your suggestions.

**Comment by lpy-neo** [comment]
Does anybody know how to make this work for YCM in vim? I've tried to update the code of gopls (in YCM) to master, but the completion of unimported packages still doesn't work.

**Comment by 110y** [comment]
When we're using `go mod` and enabling `completeUnimported`, `gopls` seems to suggest all modules in module caches. As a result, too many modules appear as candidates.

I think it is helpful that `gopls` only suggests modules that the project we are editing are depending on excepting `indirect` modules, isn't it?

**Comment by stamblerre** [comment]
We haven't yet implemented ranking for unimported candidates, but I think that issue can be solved with improved rankings. The ordering should probably be something like standard library packages are preferred over dependencies in the main module followed by packages in the module cache.

We still want to suggest packages from the module cache that haven't been added as dependencies because users may still be interested in importing them.

**Comment by 110y** [comment]
@stamblerre 

Thank you for your response!

> The ordering should probably be something like standard library packages are preferred over dependencies in the main module followed by packages in the module cache.

It sounds good! 

> We still want to suggest packages from the module cache that haven't been added as dependencies because users may still be interested in importing them.

I think we have several different needs for `completeUnimported` like below:

1. do not complete unimported packages.
2. complete only standard packages.
3. 2 + direct dependent modules.
4. 3 + indirect dependent modules.
5. 4 + all module caches.

So, I think it is better that `gopls` provides the option to control `completeUnimported` behaviour (or change `completeUnimported` option from just `bool` type to string type like `hoverKind`) in addition to sorting you mentioned.
How about this?

**Comment by ianthehat** [comment]
I don't think the number of results matters, only the order in which they appear.
So long as the ranking is good, there is no need for any kind of control of filter of the nature you suggest, the editor can just limit the total number of candidates shown.

**Comment by hbd** [comment]
@110y to @ianthehat's point, why have yet another option (or n+ options), if instead the packages are ranked by some hierarchy? e.g. (autocompletions ranked first to last)
1. imported packages
2. stdlib
3. unimported packages
4. ...

**Comment by heschik** [comment]
I think the only thing left to do here is turn it on by default.

**Comment by muirdm** [comment]
What about the increased completion latency when there are a lot of unimported packages?

**Comment by zulhfreelancer** [comment]
@heschik 

> I think the only thing left to do here is turn it on by default.

How to turn it on currently?

**Comment by stamblerre** [comment]
You can add the following to your VSCode settings:

```json5
"gopls": {
    "completeUnimported": true
}
```

**Comment by zulhfreelancer** [comment]
@stamblerre thank you. Any idea why this popup keeps showing?

<img width="462" alt="Screen Shot 2019-11-26 at 1 30 37 PM" src="https://user-images.githubusercontent.com/3416976/69602004-ff027280-1050-11ea-9832-0c9ba43b6259.png">

I already have `gopls` installed by pressing that _Update_ button. I also tried cloning `github.com/golang/tools` repo, and ran `go install` inside the `gopls` folder.

The popup remains there everything I opened VS Code. Quitting doesn't help.
